TLM/TLM/Geometry/Impl/SegmentGeometry.cs

```diff
@@ -898,13 +898,29 @@ public class SegmentGeometry : IEquatable<SegmentGeometry> {
 			return contains;
 		}
 
-		/// <summary>
-		/// Determines if, according to the stored geometry data, the managed segment is only connected to highways at the given node.
-		/// 
-		/// A segment geometry verification is not performed.
-		/// </summary>
-		/// <returns>true, if, according to the stored geometry data, the managed segment is only connected to highways at the given node, false otherwise</returns>
-		public bool HasOnlyHighways(bool startNode) {
+        /// <summary>
+        /// Determines if, according to the stored geometry data, the segment is elgible for turning right-on-red.
+        /// 
+        /// A segment geometry verification is not performed.
+        /// </summary>
+        /// <returns>true, </returns>
+        public bool HasPreferredSegment() {
+            bool hasRightSegStart = HasRightSegment(true);
+            bool hasRightSegEnd = HasRightSegment(false);
+            bool hasLeftSegStart = HasLeftSegment(true);
+            bool hasLeftSegEnd = HasLeftSegment(false);
+            return (
+                ((Constants.ServiceFactory.SimulationService.LeftHandDrive ? hasLeftSegStart || hasLeftSegEnd : hasRightSegStart || hasRightSegEnd)) ||
```

Comment:
Is it possible to simplify this a little bit because I am not sure I understand what is going here 😄

Reply:
Quite frankly me either. I don't know the different between the startNode being true and false. Basically, what's happening is it's searching for if a segment has a segment to its right or to its left. If it does, it stores it in a bool, then I use those books to compare. The reason I check both the start and the end is that I've noticed it works much better than when I only do the end and/or only do the start, however I could probably do it better if I knew the difference lol

Reply:
I will try to refactor this a little bit because I think I know how it should look like. Give me some time 😉

Reply:
Ok, I think I've fixed all issues after refactoring.

Reply:
instead of HasPreferredSegment() add this (it has more descriptive name what it's doing):
```cs
/// <summary>
/// Determines if, according to the stored geometry data, the segment is eligible for turning right-on-red.
/// 
/// A segment geometry verification is not performed. 
/// </summary>
/// <param name="segmentEndGeometry">segment end geometry of incoming segment to node</param>
/// <returns></returns>
public bool HasValidTurnOnRedOutgoingSegment(SegmentEndGeometry segmentEndGeometry) {
    bool isLeftHandDrive = Constants.ServiceFactory.SimulationService.LeftHandDrive;
    return isLeftHandDrive ? segmentEndGeometry.NumOutgoingLeftSegments > 0 : segmentEndGeometry.NumOutgoingRightSegments > 0;
}
```
and in JunctionRestrictionsManager.cs delete lines 269 and 270 and replace with this:
```cs
segGeo?.HasValidTurnOnRedOutgoingSegment(endGeo) == true;
```

As you can see we don't have to check if its one-way, I think

Reply:
And of course lines 352 and 354 in VehicleBehaviorManager.cs 😉

Reply:
I can't tell exactly where, but from what I know it also needs to detect if it's a left-to-left from a one-way to a one-way, as vehicles may turn left on red if it's from a one-way road to another one-way road.

Reply:
Thank you for looking at that! I'm glad you found a better way to do it, it seems to be much more stable now. Apologies I wasn't able to figure this out before >.<

Reply:
Ok, I can see this now in my game..
![notitle](https://user-images.githubusercontent.com/19638970/52525932-8bee6200-2cb1-11e9-9d9c-6623737388cd.jpg)

Reply:
I'm noticing they're not actually performing the right on red anymore either. Why is it that every time we fix this, it breaks again? 😫 

I'll take another look for the 9 billionth time and maybe see if I can't make a major change to the way it actually performs it.